src/reactor/registration.rs

```diff
@@ -0,0 +1,394 @@
+use reactor::{Handle, Direction};
+
+use futures::{Async, Poll};
+use futures::task::{self, Task};
+use mio::{self, Evented};
+
+use std::{io, mem, usize};
+use std::cell::UnsafeCell;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::SeqCst;
+
+/// Handle to a reactor registration.
+///
+/// A registration represents an I/O resource registered with a Reactor such
+/// that it will receive task notifications on readiness.
+///
+/// The registration is lazily made and supports concurrent operations. This
+/// allows a `Registration` instance to be created without the reactor handle
+/// that will eventually be used to drive the resource.
+///
+/// The difficulty is due to the fact that a single registration drives two
+/// separate tasks -- A read half and a write half.
+#[derive(Debug)]
+pub struct Registration {
```

Comment:
`Registration` is introduced as the most minimal API that is required to integrate with the reactor. It is only pair of readiness streams (one for read, one for write).